src/pharmpy/tools/amd/run.py

```diff
@@ -1118,6 +1134,51 @@ def validate_input(
         validate_allometric_variable(model, allometric_variable)
 
     # COVSEARCH
+    if mechanistic_covariates:
+        allowed_covariates = get_covariates_allowed_in_covariate_effect(model)
+        allowed_parameters = allowed_parameters = set(get_pk_parameters(model)).union(
+            str(statement.symbol) for statement in model.statements.before_odes
+        )
+        for c in mechanistic_covariates:
+            if isinstance(c, str):
+                if c not in allowed_covariates:
+                    raise ValueError(
+                        f'Invalid mechanistic covariate: {c}.'
+                        f' Must be in {sorted(allowed_covariates)}'
+                    )
+            else:
+                if len(c) != 2:
+                    raise ValueError(
+                        f'Invalid argument in `mechanistic_covariate`: {c}.'
+                        f' Tuples need to be given with one parameter and one covariate.'
+                    )
+                cov_found = False
+                par_found = False
+                for a in c:
+                    if a in allowed_covariates:
+                        if cov_found:
+                            raise ValueError(
+                                f'`mechanistic_covariates` contain invalid argument: got `{c}`,'
+                                f' which contain two covariates.'
+                                f' Tuples need to be given with one parameter and one covariate.'
+                            )
+                        cov_found = True
+                    elif a in allowed_parameters:
+                        if par_found:
+                            raise ValueError(
+                                f'`mechanistic_covariates` contain invalid argument: got `{c}`,'
+                                f' which contain two parameters.'
+                                f' Tuples need to be given with one parameter and one covariate.'
+                            )
+                        par_found = True
+                    else:
+                        raise ValueError(
+                            f'`mechanistic_covariates` contain invalid argument: got `{c}`,'
+                            f' which contain {a}.'
+                            f' Tuples need to be given with one parameter and one covariate.'
+                        )
+            # Typechecking performed elsewhere?
```

Comment:
This comment should be removed right? And to answer the question, this function does the type checking:
https://github.com/pharmpy/pharmpy/blob/cb5a277c678a0a24cff2bc7586e9f9109235ad18/src/pharmpy/internals/fn/type.py#L20
It is called using a decorator on the `validate_input` function of each tool, e.g.:
https://github.com/pharmpy/pharmpy/blob/cb5a277c678a0a24cff2bc7586e9f9109235ad18/src/pharmpy/tools/modelsearch/tool.py#L294-L296